src/util/buffvec.h

```diff
@@ -0,0 +1,893 @@
+
+        buffvec_detail::move_into(buffvec, m_data, m_size);
+        buffvec_detail::destroy(m_data, m_data + m_size);
+        if(m_data != reinterpret_cast<pointer>(&m_initial_buffer)) {
+            memory::deallocate(m_data);
+        }
+
+        m_data = buffvec;
+        m_capacity = new_capacity;
+    }
+
+public:
+    constexpr buffvec() noexcept = default;
+
+    buffvec(buffvec const& other) {
+        size_type const pos = other.m_size;
+        m_size = pos;
+        if(pos <= initial_size) {
+            m_data = reinterpret_cast<pointer>(&m_initial_buffer);
+            m_capacity = initial_size;
+        } else {
+            m_data = reinterpret_cast<pointer>(memory::allocate(static_cast<std::size_t>(pos) * sizeof(value_type)));
+            m_capacity = pos;
+        }
+        buffvec_detail::copy_into(m_data, other.m_data, pos);
+    }
+
+    buffvec& operator=(buffvec const& other) {
+        if(this != &other) {
+            clear();
+            reserve(other.m_size);
+            m_size = other.m_size;
+            buffvec_detail::copy_into(m_data, other.m_data, m_size);
+        }
+        return *this;
+    }
+
+    // Depending on initial_size, move construction may still be very expensive
+    buffvec(buffvec&& other) {
+        if(other.m_data != reinterpret_cast<pointer>(&other.m_initial_buffer)) {
+            m_data = other.m_data;
+            m_size = other.m_size;
+            m_capacity = other.m_capacity;
+            other.m_data = reinterpret_cast<pointer>(&other.m_initial_buffer);
+            other.m_size = 0;
+            other.m_size = initial_size;
+        } else {
+            m_data = reinterpret_cast<pointer>(&m_initial_buffer);
+            m_size = other.m_size;
+            m_capacity = initial_size;
+            buffvec_detail::move_into(m_data, other.m_data, m_size);
+            buffvec_detail::destroy(other.begin(), other.end()); // would happen during destruction ´other´ anyway
+            other.m_size = 0;
+        }
+    }
+
+    // Depending on initial_size, move assignment may still be very expensive
+    buffvec& operator=(buffvec&& other) {
+        using std::swap;
+        if(this != &other) {
+            if(other.m_data != reinterpret_cast<pointer>(&other.m_initial_buffer)) {
+                if(m_data != reinterpret_cast<pointer>(&m_initial_buffer)) {
+                    swap(m_data, other.m_data);
+                    swap(m_size, other.m_size);
+                    swap(m_capacity, other.m_capacity);
+                } else {
+                    buffvec_detail::destroy(begin(), end());
+                    m_data = other.m_data;
+                    m_size = other.m_size;
+                    m_capacity = other.m_capacity;
+                    other.m_data = reinterpret_cast<pointer>(&other.m_initial_buffer);
+                    other.m_size = 0;
+                    other.m_size = initial_size;
```

Comment:
other.m_size is assigned twice

Reply:
Copy and paste of above code, thanks for catching it!